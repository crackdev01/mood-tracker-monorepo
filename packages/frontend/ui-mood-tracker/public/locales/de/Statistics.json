{
  "header": "Statistiken",
  "description": "Übersicht der statistischen Daten Ihrer Stimmungseinträge\n",
  "moodStatus": {
    "placeholder": "Bitte wählen Sie Ihre Stimmung",
    "relaxed": "Entspannt",
    "motivated": "Motiviert",
    "energetic": "Energisch",
    "curious": "Neugierig",
    "confident": "Zuversichtlich"
  },
  "limit": {
    "dropdown": {
      "default": "Default Limit: Last 7 entries",
      "custom": "Set custom entry"
    },
    "input": {
      "placeholder": "Enter limit"
    }
  },
  "chart": {
    "xaxis": "Entry Dates",
    "yaxis": "Intensity"
  }
}
