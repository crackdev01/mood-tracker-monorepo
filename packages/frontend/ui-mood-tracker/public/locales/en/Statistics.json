{
  "header": "Statistics",
  "description": "Overview of statistical data of your mood entries",
  "moodStatus": {
    "placeholder": "Please select your mood",
    "relaxed": "Relaxed",
    "motivated": "Motivated",
    "energetic": "Energetic",
    "curious": "Curious",
    "confident": "Confident"
  }
}
