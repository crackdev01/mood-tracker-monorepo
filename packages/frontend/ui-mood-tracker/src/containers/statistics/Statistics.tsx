import React, { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import dayjs from 'dayjs';
import { Header, Dropdown, Input } from 'semantic-ui-react';

import LineChart from '../../components/statistics/LineChart';
import { ApplicationState } from '../../store';

import './statistics.scss';

const Statistics = () => {
  const { t } = useTranslation(['Statistics']);
  const DEFAULT_LIMIT = 7;
  const moods = useSelector((state: ApplicationState) => state.moodReducer.mood);
  const [chartData, setChartData] = useState([{ x: '', y: 0 }]);
  const [defaultLimit, setDefaultLimit] = useState(DEFAULT_LIMIT);
  const [showCustomLimit, setShowCustomLimit] = useState(false);

  const limiterOptions = [
    {
      key: 0,
      value: false,
      text: t('limit.dropdown.default'),
    },
    {
      key: 1,
      value: true,
      text: t('limit.dropdown.custom'),
    },
  ];

  const showLimiter = (_: any, data: any) => {
    setShowCustomLimit(data.value);
    if (!data.value) setDefaultLimit(DEFAULT_LIMIT);
  };

  const updateLimiter = (_: any, data: any) => {
    setDefaultLimit(data.value < DEFAULT_LIMIT ? DEFAULT_LIMIT : data.value);
  };

  const sortMoodEntriesByDate = moods.sort((a: any, b: any) => {
    const aDate = new Date(a.mood_enteredAt + ' ' + a.mood_enteredAt);
    const bDate = new Date(b.mood_enteredAt + ' ' + b.mood_enteredAt);

    return bDate.getTime() - aDate.getTime();
  });

  const constructData = () => {
    const intensityEntries = sortMoodEntriesByDate
      .map((m: any) => {
        return {
          x: dayjs(m.mood_enteredAt).format('DD.MM.YY'),
          y: m.mood_intensity,
        };
      })
      .slice(0, defaultLimit);
    setChartData(intensityEntries);
  };

  useEffect(() => {
    constructData();
  }, [defaultLimit]);

  return (
    <article className="statistics">
      <section className="statistics__header">
        <Header as="h2">{t('header')}</Header>

        <Header as="h3">{t('description')}</Header>
      </section>

      <section className="statistics__limit-options">
        <Dropdown
          placeholder={t('limit.dropdown.default')}
          fluid
          selection
          onChange={showLimiter}
          options={limiterOptions}
        />

        {showCustomLimit && (
          <Input placeholder={t('limit.input.placeholder')} onChange={updateLimiter} />
        )}
      </section>

      <section className="statistics__graph">
        <LineChart data={chartData} title="Mood Entries" />
      </section>
    </article>
  );
};

export default Statistics;
