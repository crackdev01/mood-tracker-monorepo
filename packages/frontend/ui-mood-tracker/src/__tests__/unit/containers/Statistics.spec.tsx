import React from 'react';
import { useSelector } from 'react-redux';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Statistics from '../../../containers/statistics/Statistics';
import { mockMoods } from '../../fixtures';

Enzyme.configure({ adapter: new Adapter() });

const useSelectorMock = useSelector as jest.Mock<any>;
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

describe('Statistics', () => {
  beforeEach(
    useSelectorMock.mockImplementationOnce((s) => s({ moodReducer: { mood: mockMoods } }))
  );

  test('renders default', () => {
    const w = shallow(<Statistics />);
    expect(w.debug()).toMatchSnapshot();
  });

  test('renders input field on changing to custom limiter value', () => {
    const w = shallow(<Statistics />);
    w.find('Dropdown').at(0).simulate('change', undefined, { value: 1 });
    const inputField = w.find('Input').at(0);
    expect(inputField.exists()).toBeTruthy();
    inputField.simulate('change', undefined, { value: 10 });
    expect(w.debug()).toMatchSnapshot();
  });
});
