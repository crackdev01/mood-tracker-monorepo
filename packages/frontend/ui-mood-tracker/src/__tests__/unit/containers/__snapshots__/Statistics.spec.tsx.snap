// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Statistics renders default 1`] = `
"<article className=\\"statistics\\">
  <section className=\\"statistics__header\\">
    <Header as=\\"h2\\">
      header
    </Header>
    <Header as=\\"h3\\">
      description
    </Header>
  </section>
  <section className=\\"statistics__limit-options\\">
    <Dropdown placeholder=\\"limit.dropdown.default\\" fluid={true} selection={true} onChange={[Function: showLimiter]} options={{...}} additionLabel=\\"Add \\" additionPosition=\\"top\\" closeOnBlur={true} closeOnEscape={true} deburr={false} icon=\\"dropdown\\" minCharacters={1} noResultsMessage=\\"No results found.\\" openOnFocus={true} renderLabel={[Function: renderItemContent]} searchInput=\\"text\\" selectOnBlur={true} selectOnNavigation={true} wrapSelection={true} />
  </section>
  <section className=\\"statistics__graph\\">
    <LineChart data={{...}} title=\\"Mood Entries\\" />
  </section>
</article>"
`;

exports[`Statistics renders input field on changing to custom limiter value 1`] = `
"<article className=\\"statistics\\">
  <section className=\\"statistics__header\\">
    <Header as=\\"h2\\">
      header
    </Header>
    <Header as=\\"h3\\">
      description
    </Header>
  </section>
  <section className=\\"statistics__limit-options\\">
    <Dropdown placeholder=\\"limit.dropdown.default\\" fluid={true} selection={true} onChange={[Function: showLimiter]} options={{...}} additionLabel=\\"Add \\" additionPosition=\\"top\\" closeOnBlur={true} closeOnEscape={true} deburr={false} icon=\\"dropdown\\" minCharacters={1} noResultsMessage=\\"No results found.\\" openOnFocus={true} renderLabel={[Function: renderItemContent]} searchInput=\\"text\\" selectOnBlur={true} selectOnNavigation={true} wrapSelection={true} />
    <Input placeholder=\\"limit.input.placeholder\\" onChange={[Function: updateLimiter]} type=\\"text\\" />
  </section>
  <section className=\\"statistics__graph\\">
    <LineChart data={{...}} title=\\"Mood Entries\\" />
  </section>
</article>"
`;
